apache/zeppelin
# [ZEPPELIN-4905] Kotlin interpreter new API

### What is this PR for?
This PR provides Kotlin interpreter users with the new version of scripting compiler (1.4) which results in following:
- Users may use new features of Kotlin 1.4.
- Compile times are significantly reduced for big notebooks: some performance issues in scripting compiler resolution were fixed.
- Completion feature is now natively supported by REPL compiler, and it's a way more smart.
- Interpreter itself and notebook scripts dependencies were splitted into different submodules.
- Implicit receivers logic was transformed to a new base-class logic.

### What type of PR is it?
Improvement

### Todos
- [ ] - Check and fix tests for Kotlin Spark interpreter

### What is the Jira issue?
https://issues.apache.org/jira/projects/ZEPPELIN/issues/ZEPPELIN-4905

### How should this be tested?
* Travis CI build: https://travis-ci.org/github/ileasile/zeppelin/builds/705291140
* Check unit tests for Kotlin and KotlinSpark interpreters
* Check that script dependencies specified for Kotlin interpreter are availible in notebook cells

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update?
**No.**
* Is there breaking changes for older versions?
**Yes.** Kotlin itself has breaking changes in 1.4: https://youtrack.jetbrains.com/issues/KT?q=Tag:%20language-committee-approved%20Target%20versions:%201.4-M1,%201.4-M2,%201.4-M3,%201.4.0%20%23Fixed%20&_ga=2.173677219.12913552.1593942860-2018140051.1491349376
Also, some minor changes were made in scripting REPL itself:
  * `InvokeWrapper` was moved to another package (`org.apache.zeppelin.kotlin.script.InvokeWrapper`) which may break the notebooks which use it
  * Internal variables like `kc` are now not shown in the list of all variables.
* Does this needs documentation?
**No.** Existing documentation is enough.
